feat: Add `RoundMode` (half even; half away from zero) and fix inconsistent rounding between float and decimal

Julian-J-S opened the pull request:
fix #21800 (though as @orlp mentioned there are theoretically more round modes to add)

# Current Problem ⚠️ 
- current `round` expression has no `mode` and is inconsistent 
  - `Decimal`: uses "half to even"
  - `Float`: uses "half away from zero"

# PR Solution 🍀 
- implement missing round modes
  - `Decimal`: implement "half away from zero"
  - `Float`: implement "half to even"
- set "half to even" as default for float & decimal

## Example
```python
df = pl.DataFrame(
    {
        "f64": [-3.5, -2.5, -1.5, -0.5, 0.5, 1.5, 2.5, 3.5],
        "d": ["-3.5", "-2.5", "-1.5", "-0.5", "0.5", "1.5", "2.5", "3.5"],
    },
    schema={
        "f64": pl.Float64,
        "d": pl.Decimal(scale=1),
    },
)

df.with_columns(
    pl.all().round().name.suffix("_default"),
    pl.all().round(mode="half_away_from_zero").name.suffix("_away_from_zero"),
    pl.all().round(mode="half_to_even").name.suffix("_to_even"),
)

shape: (8, 8)
┌──────┬──────────────┬─────────────┬──────────────┬────────────────────┬──────────────────┬─────────────┬──────────────┐
│ f64  ┆ d            ┆ f64_default ┆ d_default    ┆ f64_away_from_zero ┆ d_away_from_zero ┆ f64_to_even ┆ d_to_even    │
│ ---  ┆ ---          ┆ ---         ┆ ---          ┆ ---                ┆ ---              ┆ ---         ┆ ---          │
│ f64  ┆ decimal[*,1] ┆ f64         ┆ decimal[*,1] ┆ f64                ┆ decimal[*,1]     ┆ f64         ┆ decimal[*,1] │
╞══════╪══════════════╪═════════════╪══════════════╪════════════════════╪══════════════════╪═════════════╪══════════════╡
│ -3.5 ┆ -3.5         ┆ -4.0        ┆ -4.0         ┆ -4.0               ┆ -4.0             ┆ -4.0        ┆ -4.0         │
│ -2.5 ┆ -2.5         ┆ -2.0        ┆ -2.0         ┆ -3.0               ┆ -3.0             ┆ -2.0        ┆ -2.0         │
│ -1.5 ┆ -1.5         ┆ -2.0        ┆ -2.0         ┆ -2.0               ┆ -2.0             ┆ -2.0        ┆ -2.0         │
│ -0.5 ┆ -0.5         ┆ -0.0        ┆ 0.0          ┆ -1.0               ┆ -1.0             ┆ -0.0        ┆ 0.0          │
│ 0.5  ┆ 0.5          ┆ 0.0         ┆ 0.0          ┆ 1.0                ┆ 1.0              ┆ 0.0         ┆ 0.0          │
│ 1.5  ┆ 1.5          ┆ 2.0         ┆ 2.0          ┆ 2.0                ┆ 2.0              ┆ 2.0         ┆ 2.0          │
│ 2.5  ┆ 2.5          ┆ 2.0         ┆ 2.0          ┆ 3.0                ┆ 3.0              ┆ 2.0         ┆ 2.0          │
│ 3.5  ┆ 3.5          ┆ 4.0         ┆ 4.0          ┆ 4.0                ┆ 4.0              ┆ 4.0         ┆ 4.0          │
└──────┴──────────────┴─────────────┴──────────────┴────────────────────┴──────────────────┴─────────────┴──────────────┘
```

# Open
- I will just quote @orlp:
> Thus I'd suggest the following names (and their half_ cousins):
to_even, to nearest even (default),
to_zero, to 0,
away_from_zero, away from 0 to +inf for positive, -inf for negative,
ceil, to +inf,
floor, to -inf,
stochastic, as explained earlier.

IMO having consistent rounding across types and the two most used round modes will already be a great benefit and cover 99% of use-cases.
We can implement the other modes in the future if time and demand is there.

orlp commented:
I think we need to do another pass over the names. I think `half_away_from_zero` is a bit too wordy, and I also just realized my usage of `'up'` and `'down'` conflicts with both [Python](https://docs.python.org/3/library/decimal.html#rounding-modes) and [Java](https://docs.oracle.com/javase/8/docs/api/java/math/RoundingMode.html#enum.constant.summary).

Julian-J-S commented:
@orlp as discussed I adjusted the names and implemented
- `half_to_even` (was missing for f32/f64)
- `half_away_from_zero` (was missing for Decimal)
- set `half_to_even` as default for both

please have a look 🤓

wence- commented on crates/polars-python/src/lazyframe/visitor/expr_nodes.rs:
```diff
@@ -1272,7 +1286,15 @@ pub(crate) fn into_py(py: Python<'_>, expr: &AExpr) -> PyResult<PyObject> {
+                FunctionExpr::Round { decimals, mode } => (
+                    "round",
+                    decimals,
+                    match mode {
+                        RoundMode::HalfToEven => "half_to_even",
+                        RoundMode::HalfAwayFromZero => "half_away_from_zero",
+                    },
```
Since the enum is annotated with `strum`, we can just use:
```
Into::<&str>::into(mode)
```
here, I think.

wence- replied:
Could you please bump the minor IR version in `visit.rs` (in the impl of `NodeTraverser`).

wence- commented on crates/polars-python/src/lazyframe/visitor/expr_nodes.rs:
```diff
@@ -402,6 +403,19 @@ impl PyWindowMapping {
     }
 }
 
+#[pyclass(name = "RoundMode")]
+pub struct PyRoundMode {
+    inner: RoundMode,
+}
+
+#[pymethods]
+impl PyRoundMode {
+    #[getter]
+    fn kind(&self) -> &str {
+        self.inner.into()
+    }
+}
+
```
Then there is no need for these.

orlp commented on crates/polars-ops/src/series/ops/round.rs:
```diff
@@ -1,43 +1,92 @@
+                    } else {
+                        // Note we do the computation on f64 floats to not lose precision
+                        // when the computation is done, we cast to f32
+                        let multiplier = 10.0.pow(decimals as f64);
```
You can use `powi` without casting the exponent.

orlp commented on crates/polars-ops/src/series/ops/round.rs:
```diff
@@ -48,27 +97,34 @@ pub trait RoundSeries: SeriesSealed {
+            let res = match mode {
+                RoundMode::HalfToEven => ca.apply_values(|v| {
                     let rem = v % multiplier;
                     let is_v_floor_even = ((v - rem) / multiplier) % 2 == 0;
```
Can you please avoid the extra division by doing a modulo with `2 * multiplier` and some extra logic? `i128` division is very very expensive.
